src/librustc_infer/infer/error_reporting/mod.rs

```diff
@@ -1682,49 +1682,70 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
-                            let sp = if has_bounds && !is_impl_trait {
-                                sp.to(self
-                                    .tcx
-                                    .sess
-                                    .source_map()
-                                    .next_point(self.tcx.sess.source_map().next_point(sp)))
-                            } else {
-                                sp
-                            };
-                            (sp, has_bounds, is_impl_trait)
-                        })
-                    } else {
-                        None
-                    }
-                })
+        let generics = self
+            .in_progress_tables
+            .and_then(|table| table.borrow().hir_owner)
+            .map(|table_owner| self.tcx.generics_of(table_owner.to_def_id()));
+        let type_param_span = match (generics, bound_kind) {
+            (Some(ref generics), GenericKind::Param(ref param)) => {
+                // Account for the case where `param` corresponds to `Self`,
+                // which doesn't have the expected type argument.
+                if !(generics.has_self && param.index == 0) {
+                    let type_param = generics.type_param(param, self.tcx);
+                    type_param.def_id.as_local().map(|def_id| {
+                        // Get the `hir::Param` to verify whether it already has any bounds.
+                        // We do this to avoid suggesting code that ends up as `T: 'a'b`,
+                        // instead we suggest `T: 'a + 'b` in that case.
+                        let id = hir.as_local_hir_id(def_id);
+                        let mut has_bounds = false;
+                        if let Node::GenericParam(param) = hir.get(id) {
+                            has_bounds = !param.bounds.is_empty();
+                        }
+                        let sp = hir.span(id);
+                        // `sp` only covers `T`, change it so that it covers
+                        // `T:` when appropriate
+                        let is_impl_trait = bound_kind.to_string().starts_with("impl ");
+                        let sp = if has_bounds && !is_impl_trait {
+                            sp.to(self
+                                .tcx
+                                .sess
+                                .source_map()
+                                .next_point(self.tcx.sess.source_map().next_point(sp)))
+                        } else {
+                            sp
+                        };
+                        (sp, has_bounds, is_impl_trait)
+                    })
+                } else {
+                    None
+                }
             }
             _ => None,
         };
+        let new_lt = generics
+            .as_ref()
+            .and_then(|g| {
+                let possible = ["'a", "'b", "'c", "'d", "'e", "'f", "'g", "'h", "'i", "'j", "'k"];
+                let lts_names = g
+                    .params
```

Comment:
Hmm I think maybe this won't work if the code is in an impl?

```rust
impl<'a> Foo {
    fn bar(&self) -> Box<dyn Debug> {
        // the "generics" of bar don't include `'a` directly, have to go to the parent def-id
```

maybe we want a helper method for this kind of search?

Reply:
True. Inconvenient 😩

Reply:
I tried to cause the same E0311 that would trigger the suggestion, instead getting

```
error[E0623]: lifetime mismatch
  --> file8.rs:17:49
   |
17 |     fn bat<G>(&self, g: G, dest: &mut usize) -> impl FnOnce() + '_
   |                                  ----------     ^^^^^^^^^^^^^^^^^^
   |                                  |              |
   |                                  |              ...but data from `dest` is returned here
   |                                  this parameter and the return type are declared with different lifetimes...
```

which could do with an improvement regardless.

Reply:
Ok, managed to trigger it (without the `self` elided lifetime):

```
error[E0311]: the parameter type `G` may not live long enough
  --> file8.rs:17:42
   |
17 |     fn bat<G>(g: G, dest: &mut usize) -> impl FnOnce() + '_
   |                                          ^^^^^^^^^^^^^^^^^^
   |
note: the parameter type `G` must be valid for the anonymous lifetime #1 defined on the method body at 17:5...
  --> file8.rs:17:5
   |
17 | /     fn bat<G>(g: G, dest: &mut usize) -> impl FnOnce() + '_
18 | |     where G: Get<usize> {
19 | |         move || {
20 | |             *dest = g.get();
21 | |         }
22 | |     }
   | |_____^
note: ...so that the type `[closure@file8.rs:19:9: 21:10 g:G, dest:&mut usize]` will meet its required lifetime bounds
  --> file8.rs:17:42
   |
17 |     fn bat<G>(g: G, dest: &mut usize) -> impl FnOnce() + '_
   |                                          ^^^^^^^^^^^^^^^^^^
help: consider introducing an explicit lifetime bound
   |
17 |     fn bat<'b, G: 'b>(g: G, dest: &mut usize) -> impl FnOnce() + '_ + 'b
   |            ^^^^^                                                    ^^^^
```

<img width="846" alt="Screen Shot 2020-05-27 at 2 12 18 PM" src="https://user-images.githubusercontent.com/1606434/83072751-2c41d480-a024-11ea-83d3-94d330b5f928.png">

Note that this is with a commit I haven't pushed yet.